### PR TITLE
Initial support for reactive bulk operations.

### PR DESCRIPTION
**Pull request by christophstrobl**
Closes: #2821

### CODE REVIEW

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+							List<Mono<Object>> monos = it.stream().map(saved -> {
+								return maybeInvokeAfterSaveCallback(saved);
+							}).collect(Collectors.toList());
+
+							return Flux.concat(monos).then(Mono.just(state));
+						});
+			});
+		} catch (RuntimeException ex) {
+
+			if (ex instanceof MongoBulkWriteException) {
```
Comment (mp911de) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java:257]: Can these even happen? Why not `catch(MongoBulkWriteException e)`?

Comment (christophstrobl) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java:257]: if the operation is `ordered`, yes.
https://www.mongodb.com/docs/manual/reference/method/db.collection.bulkWrite/#error-handling

Comment (mp911de) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java:257]: But in reactive, errors are emitted as signals through the publisher (that's why I was wondering) and not through the synchronous calls and therefore, the error signal translation is missing.

Comment (christophstrobl) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java:257]: doh - you're right

Comment (mp911de) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java:257]: Okay, thanks for confirming. I'm happy to handle that aspect during the merge.

**Pull request closed without merging**